File: Striver/DP/Q45.py
# longest string chain

def longestStrChain(words):
    words.sort(key=len)
    n = len(words)
    dp = [1]*n
    mx = 1
    for i in range(1, n):
        for j in range(i):
            if len(words[i]) - len(words[j]) == 1:
                if isPredecessor(words[j], words[i]):
                    dp[i] = max(dp[i], dp[j]+1)
                    mx = max(mx, dp[i])

    return mx

def isPredecessor(s1, s2):
    i = 0
    j = 0
    while i < len(s1) and j < len(s2):
        if s1[i] == s2[j]:
            i += 1
            j += 1
        else:
            j += 1
    return i == len(s1) and j <= len(s2)

words = ["a","b","ba","bca","bda","bdca"]
print(longestStrChain(words))

File: Striver/DP/Q45.java
package Striver.DP;

import java.util.Arrays;

public class Q45 {
    // longest string chain length

    public int longestStringChain(String[] words) {
        int n = words.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        int max = 1;
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                if(isPredecessor(words[j], words[i])){
                    dp[i] = Math.max(dp[i], dp[j]+1);
                }
            }
            max = Math.max(max, dp[i]);
        }
        return max;
    }

    public boolean isPredecessor(String s1, String s2){
        if(s1.length() + 1 != s2.length()) return false;
        int i = 0, j = 0;
        while(i < s1.length() && j < s2.length()){
            if(s1.charAt(i) == s2.charAt(j)){
                i++;
                j++;
            }else{
                j++;
            }
        }
        return i == s1.length() && j == s2.length();
    }
}
